src/devices/controller/controller.cpp

```diff
@@ -1,35 +1,53 @@
+#include <Arduino.h>
 #include <FastLED.h>
+#include <arduino-timer.h>
+#include <median-filter.h>
 
-#include "Arduino.h"
 #include "FakeLedManager.hpp"
 #include "analog-button.h"
-#include "arduino-timer.h"
 #include "arduino/RadioHeadRadio.hpp"
+#include "buttons.h"
 #include "generic/NetworkManager.hpp"
 #include "generic/RadioStateMachine.hpp"
+#include "leds.h"
+
+// Which mode of control the device is in
+enum class ControllerMode {
+  // Choose an effect and a color palette.
+  Effect,
+
+  // Left/right buttons immediately make everything light up a single color.
+  DirectColor,
+
+  // Left/right buttons immediately change the palette for the current effect.
+  DirectPalette,
+};
+
+ControllerMode mode = ControllerMode::Effect;
+ControllerMode prev_mode = mode;
 
 // Pin definitions
 constexpr int kSwitchLeft = PA0;
 constexpr int kSwitchRight = PA1;
 constexpr int kSwitchBottom = PA2;
 
 constexpr int kVbattDiv = PA3;
+constexpr int kModeSwitch = PB1;
 
 constexpr int kNeopixelPin = PB15;
 
-constexpr uint16_t kLedCount = 42;
-CRGB leds[kLedCount];
-
-// LED indices
-constexpr uint8_t kStatusLeft = 39;
-constexpr uint8_t kStatusMiddle = 38;
-constexpr uint8_t kStatusRight = 37;
-
-const StripDescription kRowStrip = StripDescription(/*led_count=*/12, {Bright});
+const StripDescription kRowStrip =
+    StripDescription(/*led_count=*/12, {Bright, Controller});
 const DeviceDescription kRowDescription(2000, {kRowStrip});
+const StripDescription kPaletteStrip =
+    StripDescription(/*led_count=*/5, {Bright, Controller});
 
 constexpr uint8_t kSetEffectDelay = 60;
 
+static constexpr uint16_t kMode1Threshold = 346 / 2;
```

Comment:
I'm confused by these magic numbers, what do they represent? Trial and error?

Reply:
Added a comment - it comes from a calculator.l